### PR TITLE
[Merged by Bors] - feat(CategoryTheory): Special equalizers associated to a comonad

### PR DESCRIPTION
**Pull request by mckoen**
Given a comonad, shows that any coalgebra is an equalizer (in the category of coalgebras) of cofree coalgebras, and shows this equalizer is coreflexive. 

Also constructs the Beck fork in the category of coalgebras, and shows the fork is coreflexive as well as a split equalizer (in particular, an equalizer).

This dualises everything in [Mathlib.CategoryTheory.Monad.Coequalizer](https://leanprover-community.github.io/mathlib4_docs/Mathlib/CategoryTheory/Monad/Coequalizer.html).

This contribution was created as part of the AIM workshop "Formalizing Algebraic Geometry" in June 2024.

---

- [x] depends on: #14170 

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by kbuzzard** [comment]
Can you add a comment in the first message of this PR saying "This contribution was created as part of the AIM workshop "Formalizing Algebraic Geometry" in June 2024."?

File: Mathlib/CategoryTheory/Monad/Equalizer.lean [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+Show that any coalgebra is an equalizer of cofree coalgebras.
+-/
+
+
+/-- The top map in the equalizer diagram we will construct. -/
+@[simps!]
+def CofreeEqualizer.topMap :  (Comonad.cofree T).obj X.A ⟶ (Comonad.cofree T).obj (T.obj X.A) :=
+  (Comonad.cofree T).map X.a
+
+/-- The bottom map in the coequalizer diagram we will construct. -/
```
Comment (dagurtomas) [inline Mathlib/CategoryTheory/Monad/Equalizer.lean:47]:
```suggestion
/-- The bottom map in the equalizer diagram we will construct. -/
```

File: Mathlib/CategoryTheory/Monad/Equalizer.lean [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+/-- Construct the Beck fork in the category of coalgebras. This fork is coreflexive as well as an
+equalizer.
+-/
+@[simps!]
+def beckCoalgebraFork : Fork (CofreeEqualizer.topMap X) (CofreeEqualizer.bottomMap X) :=
+  Fork.ofι _ (CofreeEqualizer.condition X)
+
+/-- The cofork constructed is a colimit. This shows that any algebra is a (reflexive) coequalizer of
+free algebras.
+-/
```
Comment (dagurtomas) [inline Mathlib/CategoryTheory/Monad/Equalizer.lean:82]:
```suggestion
/-- The fork constructed is a limit. This shows that any coalgebra is a (coreflexive) equalizer of
cofree coalgebras.
-/
```

**Review by dagurtomas — APPROVED** [review]

File: Mathlib/CategoryTheory/Monad/Equalizer.lean (hunk 2 above)
Comment (mckoen) [reply Mathlib/CategoryTheory/Monad/Equalizer.lean:82]: Thanks Dagur!

**Comment by jcommelin** [comment]
Thanks :tada:

bors merge

**Pull request closed without merging**